### PR TITLE
Implemented solution for longest substring with K unique characters

### PR DESCRIPTION
**Pull request by haleynpatel**

### CODE REVIEW

File: solution.py [hunk 1]
```diff
@@ -1,3 +1,42 @@
...
+        while len(unique_chars) > k:
+            unique_chars[s[left]] -= 1
+            if unique_chars[s[left]] == 0:
+                del unique_chars[s[left]]
+            left += 1
+        max_length = max(max_length, right - left + 1)
+        right += 1
+    return max_length
+
+def main():
```
Comment (jpai-15) [inline solution.py:32]: I like that you put abstracted your code into functions because it makes it easy to read and go back to edit if necessary.

File: solution.py [hunk 2]
```diff
@@ -1,3 +1,42 @@
...
+    # Test cases
+    print("Longest substring with 1 unique characters in 'aabbcc' is:", longest_substring_k_unique("aabbcc", 1))  # Expected: 2
```
Comment (jpai-15) [inline solution.py:34]: Good test cases here! Maybe add a comment for the reason for each test case -- for example, test case 1 may test the edge case of having a string of 10^6 in length.

File: solution.py [hunk 3]
```diff
@@ -1,3 +1,42 @@
...
+
+    max_length = 0
+    left = 0
+    right = 0
+    unique_chars = {}
+    while right < len(s):
+        if s[right] in unique_chars:
+            unique_chars[s[right]] += 1
+        else:
+            unique_chars[s[right]] = 1
```
Comment (jpai-15) [inline solution.py:22]: I understand the logic here, but to make the code more readable, you should add comments on how the function works. It could be helpful when reviewing this code again in the future.

File: solution.py [hunk 4]
```diff
@@ -1,3 +1,42 @@
 def longest_substring_k_unique(s: str, k: int) -> int:
-    # Write your code here
-    pass
+    # Find the longest substring of a given string that contains exactly K unique characters. If no such string exists, return 0
+
+    # Edge cases 
+    if k == 0 or not s: # If k is 0 or the string is empty,
+        return 0
+    if k > len(set(s)):
+        print(f"Invalid input for k = {k}, k must be less than or equal to the number of unique characters in the string.")
```
Comment (jpai-15) [inline solution.py:8]: nice edge cases and error handling here!
